src/docmanager/context.ts

```diff
@@ -0,0 +1,456 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+'use strict';
+
+import {
+  IKernelId, IKernel, IKernelSpecIds, IContentsManager,
+  INotebookSessionManager, INotebookSession, ISessionId,
+  IContentsOpts, ISessionOptions
+} from 'jupyter-js-services';
+
+import * as utils
+  from 'jupyter-js-utils';
+
+import {
+  IDisposable, DisposableDelegate
+} from 'phosphor-disposable';
+
+import {
+  Property
+} from 'phosphor-properties';
+
+import {
+  ISignal, Signal
+} from 'phosphor-signaling';
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+import {
+  IWidgetOpener
+} from '../filebrowser/browser';
+
+import {
+  IDocumentContext, IDocumentModel, IModelFactoryOptions
+} from './index';
+
+
+/**
+ * An implementation of a document context.
+ */
+class Context implements IDocumentContext {
+  /**
+   * Construct a new document context.
+   */
+  constructor(manager: ContextManager) {
+    this._manager = manager;
+    this._id = utils.uuid();
+  }
+
+  /**
+   * A signal emitted when the kernel changes.
+   */
+  get kernelChanged(): ISignal<IDocumentContext, IKernel> {
+    return Private.kernelChangedSignal.bind(this);
+  }
+
+  /**
+   * A signal emitted when the path changes.
+   */
+  get pathChanged(): ISignal<IDocumentContext, string> {
+    return Private.pathChangedSignal.bind(this);
+  }
+
+  /**
+   * A signal emitted when the model is saved or reverted.
+   */
+  get dirtyCleared(): ISignal<IDocumentContext, void> {
+    return Private.dirtyClearedSignal.bind(this);
+  }
+
+  /**
+   * The unique id of the context.
+   */
+  get id(): string {
+    return this._id;
+  }
+  set id(value: string) {
```

Comment:
When do we want to set the id of a context? We want the id to be read-only, right?